### PR TITLE
add advanced mc 

### PR DESCRIPTION
**Pull request by mswwilson**
Not urgent to merge soon. I just wanted to get these old codes back in before they disappear. 

The new files in examples/ do not need to stay - they are just to show the usage

### CODE REVIEW

**Comment by kbarros** [comment]
This looks good to me, thanks Matt. The only suggestion I have is to consider more descriptive names, since these will be going into the public API. My first thought is `WangLandau` and `ParallelWangLandau` (the latter would be REWL I assume). This would make it easier for people to search google and learn more. What do we lose? (i.e. is there something special in MUCA that's not captured by the above names?) cc @yingwaili .

**Comment by mswwilson** [comment]
Thanks. Yeah, I struggled with what to name these types. GenEnsemble stands for 'Generalized Ensemble', which may be too obscure. These codes can be used to perform either Wang-Landau or multicanonical sampling. Wang-Landau converges towards a multicanonical ensemble in late iterations, but I know the (small) community can be picky about the distinction because multicanonical sampling actually satisfies detailed balance.
 
I originally had them as 'Multicanonical' and 'ParallelMulticanonical', and also considered something like 'FlatHistogramSampler', although the histogram criterion doesn't have to be chosen as flat. It may be more appropriate to call Wang-Landau a type of Multicanonical sampling rather than the other way around, but I'm open to any alternatives.

File: src/MonteCarlo/ParallelTempering.jl [hunk 1]
```diff
@@ -75,6 +72,8 @@ function replica_exchange!(PT::ParallelTempering, exch_start::Int64)
     end
 end
 
+function sample! end
```
Comment (yingwaili) [inline src/MonteCarlo/ParallelTempering.jl:75]: What is the purpose of this empty function?

Comment (mswwilson) [reply src/MonteCarlo/ParallelTempering.jl:75]: Oh, I can get rid of that when I make the next commit to rename things.

Comment (kbarros) [reply src/MonteCarlo/ParallelTempering.jl:75]: Something similar happens with the `step!` function -- in that context, the purpose is to create a place to store the doc string. Maybe do the same here?

**Comment by kbarros** [comment]
Ying Wai is in the process of preparing review. One thing we noticed is that Wang-Landau requires a proposal function, rather than a general sampler. It would be nice to use the latter instead, e.g., to enable the use of global moves (like a Langevin step) within Wang-Landau. Unfortunately, the `step!` function for `Langevin` provides no facility to "undo" a move. To enable this, we would need to store the trial update in a separate state array. This, in turn, would require creating some new data structures. I created https://github.com/SunnySuite/Sunny.jl/issues/76 to discuss this.

**Comment by ddahlbom** [comment]
An `extra` subdirectory sounds like a good solution. I agree the top level is just for Literate files that go on the docs. (Probably we should change `longer_examples` to something more explicit, since this directory is exclusively for the literate files that are used to make notebooks for the SunnyTutorials repo. Maybe 'longer_examples` -> 'sunny_tutorials'.)

In any case, it's exciting to see all this MC functionality going into Sunny!

**Comment by kbarros** [comment]
OK, so we would have:
```
examples/
     *.jl   #  polished Literate files, users should look at them first
     longer_tutorials/*.jl  # used to generate polished notebooks for SunnyTutorials
     extra/*.jl  # miscellaneous stuff can land here if it's not polished yet
```

**Comment by kbarros** [comment]
There is high demand for these features, so I'd like to get this merged. Remaining tasks:

* Rename `GenEnsemble -> WangLandau`
* Rename `ParallelGenEnsemble -> ParallelWangLandau`
* Move the new examples into `examples/extra/`

Other polish can be added later. @yingwaili do you agree?

Kipton

**Comment by mswwilson** [comment]
Ok, I made those changes. I'm going out for a bit, but can add a doc string to sample! later

**Review by yingwaili — COMMENTED** [review]
I am only half-way through, but would like to release my comments so @mswwilson and @kbarros can look at them and start making changes.

**Comment by kbarros** [comment]
I have merged this. We can address Ying Wai's feedback in follow up work.

File: src/MonteCarlo/GenEnsemble.jl [hunk 2]
```diff
@@ -0,0 +1,153 @@
+# data for multicanonical ensemble -- either MUCA or WL
+mutable struct GenEnsemble{PR}
```
[the remark this answers is not in the document]
Comment (kbarros) [reply src/MonteCarlo/GenEnsemble.jl]: This is a good suggestion. We will eventually need to create a `trial_step!`  function that can be accepted or rejected. See https://github.com/SunnySuite/Sunny.jl/issues/76 .

File: src/MonteCarlo/GenEnsemble.jl [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    ln_g::BinnedArray{Float64, Float64} # log density of states, ln[g(E)]
+    hist::BinnedArray{Float64, Float64} # energy histogram
+    bounds::Vector{Float64} # energy bounds to sample [E_min, E_max]
```
[the remark this answers is not in the document]
Comment (kbarros) [reply src/MonteCarlo/GenEnsemble.jl]: I will convert to Tuple

File: src/Sunny.jl [hunk 4]
```diff
@@ -102,8 +102,12 @@ include("MonteCarlo/Samplers.jl")
 include("MonteCarlo/BinnedArray.jl")
 include("MonteCarlo/ParallelTempering.jl")
 include("MonteCarlo/HistogramReweighting.jl")
+include("MonteCarlo/GenEnsemble.jl")
+include("MonteCarlo/ParallelGenEnsemble.jl")
 export propose_uniform, propose_flip, propose_delta, @mix_proposals, LocalSampler,
-    ParallelTempering, sample!
+    ParallelTempering, sample!,
```
[the remark this answers is not in the document]
Comment (kbarros) [reply src/Sunny.jl:108]: I will hide these for now since the API is subject to change

**Comment by yingwaili** [comment]
Regarding the MC examples: once they are polished, I would suggest creating a new directory dedicated to Monte Carlo methods (e.g. called `MonteCarlo` instead of `extra`) and move the examples there?

File: src/MonteCarlo/GenEnsemble.jl [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    propose::PR     # MC move proposal (function)
+    per_spin::Bool  # choose whether to use energy per spin 
+    ln_f::Float64   # factor for insstantaneous modification of ln[g(E)] 
+    E::Float64      # current energy state
+
+    function GenEnsemble(; bin_size=1.0, bounds=Float64[], propose=propose_uniform, per_spin=true, ln_f=0.0)
```
[the remark this answers is not in the document]
Comment (kbarros) [reply src/MonteCarlo/GenEnsemble.jl]: Done

File: src/MonteCarlo/GenEnsemble.jl [hunk 6]
```diff
@@ -0,0 +1,153 @@
+# data for multicanonical ensemble -- either MUCA or WL
+mutable struct GenEnsemble{PR}
+    ln_g::BinnedArray{Float64, Float64} # log density of states, ln[g(E)]
+    hist::BinnedArray{Float64, Float64} # energy histogram
+    bounds::Vector{Float64} # energy bounds to sample [E_min, E_max]
+    propose::PR     # MC move proposal (function)
+    per_spin::Bool  # choose whether to use energy per spin
```
[the remark this answers is not in the document]
Comment (kbarros) [reply src/MonteCarlo/GenEnsemble.jl]: Removed the per_spin option

File: src/MonteCarlo/GenEnsemble.jl [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    
+    Gen.ln_g[E_new] = minimum(get_vals(Gen.ln_g)) - Gen.ln_f
+    Gen.hist[E_new] = 0.0
+    reset!(Gen.hist)
+end
+
+# use WL sampling to get system in bounded energy range -- **will reset MUCA data**
+function init_system!(sys::System{N}, Gen::GenEnsemble, nsteps::Int64; limit_pad::Float64=0.0) where{N, CR}
+
+    fac = Gen.per_spin ? 1.0/length(sys.dipoles) : 1.0
```
[the remark this answers is not in the document]
Comment (kbarros) [reply src/MonteCarlo/GenEnsemble.jl]: Revised, energy is always per spin

File: src/MonteCarlo/GenEnsemble.jl [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+                end
+            end
+        end
+        Gen.ln_g[Gen.E] += 1.0
+    end
+    # initialization failed
+    return mcs
+end
+
+# perform  MUCA or WL sampling for specified number of sweeps
```
[the remark this answers is not in the document]
Comment (kbarros) [reply src/MonteCarlo/GenEnsemble.jl]: Updated

**Comment by kbarros** [comment]
This is a good idea. The current "examples" are really more like tests. This code will need to be further polished, and then we will make a proper example that can be included in the docs.

**Pull request merged**